ls1intum/Artemis
# `Exam mode`: Improve management for test exams

### Checklist
#### General
#### Server
#### Client

### Motivation and Context
With the introduction of test exams, the Exam Management Component needs to be adapted to the new requirements. 

### Description
**Exam Checklist for test exams**
For test exams, the instructor only needs to create the exercises and correctly set the points and number of exercises. 
All other tasks such as registration an generation of StudentExam is performed automatically. 
The corresponding subpages have either been adapted or hidden for test exams (see Screenshots).
Additionally, all REST-Calls have been added with a check, if they should only be used for either test exam or real exam (in conjunction with Exam Mode: Conduct test exams and refactor Course-Exam Page #5194   )

**More Structured Exam Checklist**
The Exam Checklist was split up into the tree parts, aligned to the Exam Status Component:
- Exam Preparation
- Exam Conduction
- Exam Correction

**Enhanced Exam Status Component**
The Exam Status component was adapted to the requirements of test exams. 
The Exam Correction part was hidden, the Exam Conduction part was added with the working time (for both exam types) and the number of registered students.
Additionally, the error information has been improved. If all mandatory exam preparation steps (excluding exercise preparation) are not fulfilled after the exam conduction phase starts, a warning text is displayed, as students will not be able to start the exam.

**test exams ready for production**
The admin-check in the exam-update.coponent has been removed, so instructors are now able to create test exams (in conduction with Exam Mode: Conduct test exams and refactor Course-Exam Page #5194   )
Furthermore, the number of correctionRounds for test exams has been set to 0, as only automated assessment is featured.


### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 Student

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a new test exam
4. Proceed with the test exam Checklist and navigate back to the exam list after each change and compare whether the exam status has changed accordingly.
5. Take a look at the `StudentExams` and `Students` subpages and verify, that no changes (except for the repositorys) are possible on the part of the instructor and the hint texts are displayed (see screenshot)
5. Start the test exam as a Student (or any user) and verify, that the checklist is a helpful guide for the creation of test examinations.
6. Create a new real exam 
7. Proceed with the real exam Checklist and navigate back to the exam list after each change and compare whether the exam status has changed accordingly.

**Exam Status Errors**
1. Create a test exam, wait until the start of the ExamConduction and make sure that a warning in the Exam Conduction is shown if the number of points do not match the sum of points of the exercises.
2. Create a real exam and make sure that a warning in the Exam Conduction is shown if at least one of the mandatory steps of the exam preparation (Exercises configured, Students registered and generated student exams) is not completed.

**Conduction of test exams**
You may proceed with the testing of Exam Mode: Conduct test exams and refactor Course-Exam Page #5194  , which enables the conduction of the created test exam

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage
| Class/File | Branch | Line |
|------------|-------:|-----:|
| 	ExamService.java	 | 	75%	 | 	91%	 |
| 	ExamResource.java	 | 	92%	 | 	98%	 |
| 	exam-status.component.ts	 | 	86,61%	 | 	93,54%	 |
| 	exam-update.component.ts	 | 	89,42%	 | 	94,68%	 |


### Screenshots
Enhanced exam status component
![Exam-Management-Page](https://user-images.githubusercontent.com/94070506/172166379-7f9261de-c274-4240-850c-6c08c47e1036.png)

Adapted ExamChecklist
![image](https://user-images.githubusercontent.com/94070506/172166656-bd5e6705-c553-4921-8270-9d1aaa2b610a.png)

Changed StudentExams subpage for test exams
![Hinweis-Student-Exams](https://user-images.githubusercontent.com/94070506/172166722-9374d01e-95e4-4be8-9908-177b5ac2625b.png)

Changed Students subpage for test exams
![Hinweis-Students](https://user-images.githubusercontent.com/94070506/172166735-c69919dd-ef96-4e7a-8c06-129358f44c1a.png)
